### ISSUE TITLE
Repeated Generation Regression with Qwen2-VL-7B-Instruct-4bit and default LM Studio generation config

### ISSUE THREAD

**Issue by mattjcly** [lmstudio-ai/mlx-engine#27]
### tl;dr
There has been a change in behavior between [`main` (e51ede12e1b639fd30c8797eb3bbd8b9fb3de826)](https://github.com/lmstudio-ai/mlx-engine/commit/e51ede12e1b639fd30c8797eb3bbd8b9fb3de826) and [23a41f274390ebad2b3286b19a2d677185cd3419](https://github.com/lmstudio-ai/mlx-engine/commit/23a41f274390ebad2b3286b19a2d677185cd3419), in which default LM Studio generation config results in infinite generation when it used to not.

### Overview

Repetition penalty has been observed to no longer be utilized the same way when generating with [`Qwen2-VL-7B-Instruct-4bit`](https://huggingface.co/mlx-community/Qwen2-VL-7B-Instruct-4bit) and image:
![7-Figure3-1](https://github.com/user-attachments/assets/f972d7cc-10fc-4161-a8ec-d306b7c1e9bb)

### Repro
1. Setup python environment from [`main` (e51ede12e1b639fd30c8797eb3bbd8b9fb3de826)](https://github.com/lmstudio-ai/mlx-engine/commit/e51ede12e1b639fd30c8797eb3bbd8b9fb3de826)
  ```
  python -m venv .venv
  source .venv/bin/activate
  pip install -U -r requirements.txt
  ```
2. Hardcode repetition_penalty to `1.1` in `demo.py` (default from LM Studio), temp to `0` and seed to `0` for reproducibility
  ```
  # Generate the response
  generator = create_generator(
      model_kit,
      prompt_tokens,
      None,
      images_base64,
      args.stop_strings,
      {"max_tokens": 1024, "temp": 0.0, "seed": 0, "repetition_penalty": 1.1},
  )
  ```
3. Run `demo.py` to observe repetition:
  ```
  -> % python demo.py --model ~/.cache/lm-studio/models/mlx-community/Qwen2-VL-7B-Instruct-4bit --prompt "<|im_start|>system
  You are a helpful assistant.<|im_end|>
  <|im_start|>user
  Transcribe<|vision_start|><|image_pad|><|vision_end|><|im_end|>
  <|im_start|>assistant
  " --images '/Users/matt/Downloads/7-Figure3-1.png'
  [mlx-engine] Prompt dump: <|im_start|>system
  You are a helpful assistant.<|im_end|>
  <|im_start|>user
  Transcribe<|vision_start|><|image_pad|><|vision_end|><|im_end|>
  <|im_start|>assistant
  
  [mlx-engine] Image 1: Original size (766, 836)
  [mlx-engine] Image 1: No resize needed
  Here is the transcription of the text in the image:
  
  1) a * a - b * b
  2) a * (a - b) = a * a - a * b
  3) (a - b) * (a - b) + 2 * (a - b) * b = 
     = (a - b) * (a - b) + 2 * (a * b - b * b) = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
     = a * a - 2 * a * b + b * b + 2...
  ```

Adding `"repetition_context_size": 64` does not resolve the issue

Up-ing the `repetition_penalty` to `1.2` resolves the infinite repetition, but degrades the response quality (c is never mentioned in the image):
```
Here is the transcription of your handwritten text:

1) a - b + c

2) (a - b)(c - d)

3) (a - b)^2 + 2(a - b)c

4) a^2 - ab - ac + bc
```

### Past behavior
This did not used to occur with [23a41f274390ebad2b3286b19a2d677185cd3419](https://github.com/lmstudio-ai/mlx-engine/commit/23a41f274390ebad2b3286b19a2d677185cd3419)
1. Setup environment (make sure to clean virtual environment from above first)
  ```
  git checkout 23a41f274390ebad2b3286b19a2d677185cd3419
  python -m venv .venv
  source .venv/bin/activate
  pip install -U -r requirements.txt
  ```
2. Hardcode repetition_penalty to `1.1` in `demo.py` (default from LM Studio), temp to `0` and seed to `0` for reproducibility
  ```
  # Generate the response
  generator = create_generator(model_kit, prompt_tokens, None, images_base64, 
                               {"max_tokens": 1024, 
                                "repetition_penalty":1.1, 
                                "temp": 0, 
                                "seed": 0})
 ```
3. Run `demo.py` to observe no infinite repetition:
```
-> % python demo.py --model ~/.cache/lm-studio/models/mlx-community/Qwen2-VL-7B-Instruct-4bit --prompt "<|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
Transcribe<|vision_start|><|image_pad|><|vision_end|><|im_end|>
<|im_start|>assistant
" --images '/Users/matt/Downloads/7-Figure3-1.png'
Prompt dump: <|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
Transcribe<|vision_start|><|image_pad|><|vision_end|><|im_end|>
<|im_start|>assistant

Image 1: Original size (766, 836)
Image 1: No resize needed
Here is the transcription of the text in the image:

1) a * a - b * b
2) a * (a - b) = a * a - a * b
3) (a - b) * (a - b) + 2 * (a - b) * b = 
   = (a - b) * (a - b) + 2 * a * b - b * b = 
   = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
   = a * a + 2 * a * b - b * b
4) (a - b) * (a - b) + 2 * (a - b) * b = 
   = a * a - 2 * a * b + b * b + 2 * a * b - 2 * b * b = 
   = a * a + 2 * a * b - b * b
```
**Note that if I change `repetition_penalty` to 1.0 instead of 1.1, it mirrors the infinitely repeating behavior from current main (when `main` has repetition_penalty of `1.1`). This suggests that the application of repetition penalty has changed behavior in newer` mlx-engine`**

**Comment by mattjcly** [comment]
Potentially related: https://github.com/lmstudio-ai/mlx-engine/issues/12 (although the models cited in that issue are not vision)

**Comment by Blaizzy** [comment]
@blaizzy

**Comment by Blaizzy** [comment]
Fix is merged and will be available on the next release.

**Comment by neilmehta24** [comment]
I can confirm that https://github.com/Blaizzy/mlx-vlm/pull/109 fixes this issue. Thank you!

**Comment by Blaizzy** [comment]
My pleasure!
